fish-rust/src/bitset.rs

```diff
@@ -0,0 +1,275 @@
+use num_traits::bounds::UpperBounded;
+use num_traits::{AsPrimitive, FromPrimitive, Num};
+use std::ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, Not, Shl, Shr};
+
+pub struct BitSet<T>(T);
+
+impl<
+        T: BitAnd<T, Output = T>
+            + BitOr<T, Output = T>
+            + BitAndAssign<T>
+            + BitOrAssign<T>
+            + Shl<usize, Output = T>
+            + Shr<usize, Output = T>
+            + Not<Output = T>
+            + AsPrimitive<u8>
+            + FromPrimitive
+            + UpperBounded
+            + Num
+            + AsI64
+            + AsUsize,
+    > BitSet<T>
+where
+    i64: AsPrimitive<T>,
+{
+    pub fn set(&mut self, i: usize) {
+        self.0 |= T::one() << i;
+    }
+
+    pub fn unset(&mut self, i: usize) {
+        self.0 &= !(T::one() << i);
+    }
+
+    pub fn toggle(&mut self, i: usize, v: bool) {
+        // if v { self.set(i) } else { self.unset(i) }
+        let v = T::from_u8(v as u8).unwrap();
+        let mask = T::one() << i;
+        let bit: T = ((-v.as_i64()).wrapping_shr(63)).as_();
+        self.0 = (self.0 & !mask) | (bit & mask);
```

Comment:
I think if we use xor, we don't need `bit`

	self.0 ^= T::one() << i

(untested)

Reply:
That would unconditionally flip the ith bit instead of conditionally either setting it or clearing it regardless of what the old value was, no?